src/dynamic/trace.jl

```diff
@@ -167,3 +172,39 @@ function get_submaps_shallow(choices::DynamicDSLChoiceMap)
         for (key, trie) in get_internal_nodes(choices.trie))
     Iterators.flatten((calls_iter, internal_nodes_iter))
 end
+
+## Base.getindex ##
+
+function _getindex(trace::DynamicDSLTrace, trie::Trie, addr::Pair)
+    (first, rest) = addr
+    if haskey(trie.leaf_nodes, first)
+        choice_or_call = trie.leaf_nodes[first]
+        if choice_or_call.is_choice
+            error("Unknown address $addr; random choice at $first")
+        else
+            subtrace = choice_or_call.subtrace_or_retval
+            return subtrace[rest]
+        end
+    elseif haskey(trie.internal_nodes, first)
+        return _getindex(trace, trie.internal_nodes[first], rest)
+    end
+end
+
+function _getindex(trace::DynamicDSLTrace, trie::Trie, addr)
+    if haskey(trie.leaf_nodes, addr)
+        choice_or_call = trie.leaf_nodes[addr]
+        if choice_or_call.is_choice
+            # the value of the random choice
+            return choice_or_call.subtrace_or_retval
+        else
+            # the return value of the generative function call
+            return get_retval(choice_or_call.subtrace_or_retval)
+        end
+    elseif haskey(trie.internal_nodes, addr)
+        error("No random choice or generative function call at address $addr")
+    end
+end
```

Comment:
How did this code not already exist?

Reply:
Previously, `getindex` was implemented generically for all traces in terms of `get_choices` (which returns a choice map that implements this recursive access pattern itself). For DynamicDSLTraces, which now have auxiliary state, there is now new logic required to walk the address hierarchy when accessing a particular address of this state.